File: src/sarma_cli/runtime/middleware.py
"""Runtime middleware construction for LangChain agents.

Middleware can expose powerful tools to the model. Sarma gives agents direct
file and shell access rooted at the current workspace.
"""

from __future__ import annotations

import os
import shutil
from pathlib import Path
from typing import Any


def build_agent_middleware() -> tuple[Any, ...]:
    return _build_agent_middleware(None)


def build_agent_middleware_for_model(model: Any) -> tuple[Any, ...]:
    """Build LangChain v1 agent middleware that needs a model instance."""
    return _build_agent_middleware(model)


def _build_agent_middleware(model: Any | None) -> tuple[Any, ...]:
    """Build default LangChain v1 agent middleware.

    File tools and shell commands both operate in the real current workspace.
    The filesystem backend uses virtual path mode so relative file operations
    are anchored to ``Path.cwd()`` instead of an in-memory graph state.
    """
    import warnings

    from deepagents.backends import FilesystemBackend
    from deepagents.middleware import (
        FilesystemMiddleware,
        RubricMiddleware,
        SummarizationMiddleware,
        SummarizationToolMiddleware,
    )
    from langchain.agents.middleware import (
        FilesystemFileSearchMiddleware,
        ModelRetryMiddleware,
        TodoListMiddleware,
        ToolRetryMiddleware,
    )
    from langchain.agents.middleware.shell_tool import ShellToolMiddleware

    workspace_root = Path.cwd()
    backend = FilesystemBackend(root_dir=workspace_root, virtual_mode=True)
    middleware: list[Any] = [
        TodoListMiddleware(),
        FilesystemFileSearchMiddleware(
            root_path=str(workspace_root),
            use_ripgrep=shutil.which("rg") is not None,
        ),
        FilesystemMiddleware(backend=backend),
        ShellToolMiddleware(
            workspace_root=workspace_root,
            shell_command=_shell_command(),
            tool_name="shell",
        ),
        ModelRetryMiddleware(max_retries=2),
        ToolRetryMiddleware(max_retries=2),
    ]

    if model is None:
        return tuple(middleware)

    summarization = SummarizationMiddleware(model, backend=backend)
    middleware.extend([
        summarization,
        SummarizationToolMiddleware(summarization=summarization),
    ])

    try:
        from langchain_core._api import LangChainBetaWarning
    except ImportError:  # pragma: no cover - compatibility fallback
        LangChainBetaWarning = Warning
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", LangChainBetaWarning)
        middleware.append(RubricMiddleware(model=model))

    return tuple(middleware)


def _shell_command() -> tuple[str, ...] | None:
    """Return a POSIX-compatible shell command for ShellToolMiddleware."""
    if os.name != "nt":
        return None
    candidates = [
        shutil.which("bash"),
        r"C:\Program Files\Git\bin\bash.exe",
        r"C:\Program Files\Git\usr\bin\bash.exe",
    ]
    for candidate in candidates:
        if candidate and Path(candidate).is_file():
            return (candidate,)
    return None

File: src/runtime/middleware.ts
/**
 * Runtime middleware construction for LangChain agents.
 *
 * Middleware can expose powerful tools to the model. Sarma gives agents direct
 * file and shell access rooted at the current workspace.
 *
 * JS mapping notes (vs the Python build):
 * - Python wired separate `FilesystemMiddleware`, `FilesystemFileSearchMiddleware`
 *   and `ShellToolMiddleware`. In deepagents-js a single
 *   `createFilesystemMiddleware` over an execution-capable backend
 *   (`LocalShellBackend`) exposes `ls/read_file/write_file/edit_file/glob/grep`
 *   AND an `execute` shell tool — so one middleware covers all three roles.
 * - `RubricMiddleware` and `SummarizationToolMiddleware` have no JS equivalent
 *   and are intentionally omitted; `summarizationMiddleware` covers context
 *   compaction.
 */

import {
  createMiddleware,
  todoListMiddleware,
  toolRetryMiddleware,
  summarizationMiddleware,
} from "langchain";
import { AIMessage } from "@langchain/core/messages";
import { isCommand } from "@langchain/langgraph";
import { createFilesystemMiddleware, LocalShellBackend } from "deepagents";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import type { AnyAgentMiddleware } from "langchain";

/** Build middleware that does not require a model instance. */
export function buildAgentMiddleware(): AnyAgentMiddleware[] {
  return buildMiddleware(null);
}

/** Build LangChain v1 agent middleware that needs a model instance. */
export function buildAgentMiddlewareForModel(model: BaseChatModel): AnyAgentMiddleware[] {
  return buildMiddleware(model);
}

function buildMiddleware(model: BaseChatModel | null): AnyAgentMiddleware[] {
  const workspaceRoot = process.cwd();

  // LocalShellBackend extends FilesystemBackend and adds `execute`. virtualMode
  // anchors relative file paths under the workspace (matching the Python
  // FilesystemBackend(virtual_mode=True)); shell commands run from rootDir with
  // the inherited environment.
  const backend = new LocalShellBackend({
    rootDir: workspaceRoot,
    virtualMode: true,
    inheritEnv: true,
  });

  const middleware: AnyAgentMiddleware[] = [
    todoListMiddleware(),
    createFilesystemMiddleware({ backend }) as unknown as AnyAgentMiddleware,
    sarmaModelRetryMiddleware({ maxRetries: 2 }),
    toolRetryMiddleware({
      maxRetries: 2,
      onFailure: (error: Error) => formatErrorMessage(error),
    }),
  ];

  if (model === null) {
    return middleware;
  }

  middleware.push(summarizationMiddleware({ model }) as unknown as AnyAgentMiddleware);

  return middleware;
}

export function sarmaModelRetryMiddleware(options: { maxRetries?: number } = {}): AnyAgentMiddleware {
  const maxRetries = Math.max(0, Math.trunc(options.maxRetries ?? 2));
  return createMiddleware({
    name: "modelRetryMiddleware",
    wrapModelCall: async (request, handler) => {
      let lastError: unknown = null;
      for (let attempt = 0; attempt <= maxRetries; attempt += 1) {
        try {
          return normalizeModelResponse(await handler(request));
        } catch (exc) {
          lastError = exc;
          if (attempt === maxRetries) break;
        }
      }
      throw new Error(`Model call failed after ${maxRetries + 1} attempts: ${formatErrorMessage(lastError)}`);
    },
  }) as unknown as AnyAgentMiddleware;
}

function formatErrorMessage(error: unknown): string {
  if (!(error instanceof Error)) return String(error ?? "unknown error");

  const parts = [`${error.name}: ${error.message}`];
  let cause = error.cause;
  while (cause) {
    if (cause instanceof Error) {
      parts.push(`${cause.name}: ${cause.message}`);
      cause = cause.cause;
    } else {
      parts.push(String(cause));
      break;
    }
  }
  return parts.join(" <- ");
}

// LangChain's handler may return internal response envelopes in some versions;
// only normalize shapes we know are semantically model responses.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function normalizeModelResponse(value: unknown): any {
  if (AIMessage.isInstance(value)) return value;
  if (isCommand(value)) return value;
  if (!value || typeof value !== "object") {
    throw new Error(`Unsupported model response from handler: ${typeof value}`);
  }

  const record = value as Record<string, unknown>;
  if (AIMessage.isInstance(record.message)) return record.message;
  if (Array.isArray(record.messages)) {
    const lastAi = [...record.messages].reverse().find((msg) => AIMessage.isInstance(msg));
    if (lastAi) return lastAi;
  }
  throw new Error(`Unsupported model response envelope from handler: ${Object.keys(record).join(", ") || "(empty)"}`);
}
